core/src/test/java/org/apache/iceberg/TestRemoveUnusedSpecs.java

```diff
@@ -0,0 +1,132 @@
+
+    table.updateSpec().addField("category").commit(); // 3
+    if (formatVersion == 1) {
+      table.newAppend().appendFile(newDataFile("data_bucket=0/id=5/ts=100/category=fo")).commit();
+    } else {
+      table
+          .newRowDelta()
+          .addDeletes(newDeleteFile(table.spec().specId(), "data_bucket=0/id=5/ts=100/category=fo"))
+          .commit();
+    }
+
+    table.updateSpec().addField("data").commit(); // 4
+    assertThat(table.specs()).size().as("Added specs should be present").isEqualTo(5);
+
+    PartitionSpec currentSpec = table.spec();
+    table.removeUnusedSpecs().commit();
+    assertThat(table.specs().keySet()).as("Unused specs are removed").containsExactly(1, 3, 4);
+    assertThat(table.spec()).as("Current spec shall not change").isEqualTo(currentSpec);
+  }
+
+  @TestTemplate
+  public void testRemoveUnpartitionedSpec() {
+    // clean it first to reset to un-partitioned
+    cleanupTables();
+    this.table = create(SCHEMA, PartitionSpec.unpartitioned());
+    DataFile file =
+        DataFiles.builder(table.spec())
+            .withPath("/path/to/data-0.parquet")
+            .withFileSizeInBytes(10)
+            .withRecordCount(100)
+            .build();
+    table.newAppend().appendFile(file).commit();
+    // add a bucket partition
+    table.updateSpec().addField("data_bucket", Expressions.bucket("data", 16)).commit();
+
+    // removeUnusedPartitionSpec shall not remove the un-partitioned spec
+    table.removeUnusedSpecs().commit();
+    assertThat(table.specs().keySet())
+        .as("Un-partitioned spec is still used")
+        .containsExactly(0, 1);
+
+    table.newDelete().deleteFile(file).commit();
+    DataFile bucketFile =
+        DataFiles.builder(table.spec())
+            .withPath("/path/to/data-0.parquet")
+            .withFileSizeInBytes(10)
+            .withRecordCount(100)
+            .withPartitionPath("data_bucket=0")
+            .build();
+    table.newAppend().appendFile(bucketFile).commit();
+
+    table.expireSnapshots().expireOlderThan(System.currentTimeMillis()).commit();
+    // un-partitioned spec can be removed.
+    table.removeUnusedSpecs().commit();
+    assertThat(table.specs().keySet()).as("Un-partitioned spec is removed").containsExactly(1);
+
+    // bucket spec can reset to un-partitioned
+    table.updateSpec().removeField("data_bucket").commit();
+    assertThat(table.spec().isUnpartitioned()).as("Should equal to un-partitioned").isTrue();
+    if (formatVersion == 1) {
+      assertThat(table.spec().fields().size()).as("Should have one void transform").isEqualTo(1);
+      assertThat(table.spec().specId())
+          .as("un-partitioned is evolved to use a new SpecId")
+          .isEqualTo(2);
+    } else {
+      assertThat(table.spec().fields().size()).as("Should have no fields").isEqualTo(0);
+      assertThat(table.spec().specId())
+          .as("un-partitioned is evolved to use a new SpecId")
+          .isEqualTo(2);
+    }
```

Comment:
Is it possible to simplify this by having a separate variable for expected spec ID, the only difference in the assertion is the expected spec ID which depends on the format version. The rest of the assertion is the same if I'm not missing anything.